### ISSUE TITLE
Stable Font File URLs?

### ISSUE THREAD

**Issue by davelab6** [google/fonts#1067]
@zachleat has an idea (detailed in https://github.com/zachleat/idea-book/issues/15) about how the fonts API could offer predictable font file URLs; filing this here for future discussion :)

**Comment by zachleat** [comment]
Just as background information, I was clued in that this kind of thing is happening 😇

![image](https://user-images.githubusercontent.com/39355/48258029-7b112480-e414-11e8-967f-6bfc498262e0.png)

Curious how reliable this is?

**Comment by garretrieger** [comment]
CSS responses from the API should not be copied and inlined into a pages html:
- The CSS response is specific to the user agent of the browser making the request to the CSS api. So CSS collected from one browser may not work correctly for other browsers.
- We make no guarantees that font files at older versions will always be available.

**Comment by davelab6** [comment]
Further discussion here

https://twitter.com/davelab6/status/1127289053169299459?s=19

Leaving this issue open, but we should address this better in our faq

**Comment by addyosmani** [comment]
A few thoughts on directions we could take here...

- Document that if you want to use rel=preload with stable URLs today, self-hosting may be your best option (FAQ entry). Tools like [google-webfonts-helper](https://google-webfonts-helper.herokuapp.com/fonts) are used by developers to do this today.
- If we wanted to formally support preloading, consider a "stable" font-file URL developers could leverage such as /latest.woff2 (which redirects to the appropriate version for the UA). This would probably need to factor in a few things, including caching, redirects etc.

Something like...

```
<link rel="preload" as="font" type="font/woff2" 
  href="https://fonts.gstatic.com/s/roboto/latest.woff2" crossorigin>
```

There are other options like build-time plugins, but I think these suffer from the same challenge - the URLs you statically specify in HTML to preload can easily change based on UA. We likely don't want developers having to worry about font content negotiation too much themselves.

**Comment by rsheeter** [comment]
I suppose I'm late but the suggestions here are interesting :)

Thinking "aloud," the static nature of the font files combined with dynamic css lends itself to aggressive cache durations and CDN use for the font binaries. If we had to make `/latest` dynamic (so it would send you the right font for your platform and UA) we might have to give up some of that

Another concern is that there isn't necessarily a single file to return given `unicode-range` slicing, particularly for CJK (https://www.unicodeconference.org/presentations-42/S5T3-Sheeter.pdf). Starting download earlier but pulling 10x the number of bytes requires a fast unmetered connection to be a win. Now I need:

```html
<link rel="preload" as="font" type="font/woff2" 
  href="https://fonts.gstatic.com/s/roboto/latest" crossorigin
  min-connection=(veryfast, unmetered)>
```

Our current long term play is progressive font enrichment, described at https://www.w3.org/Fonts/WG/webfonts-2018.html and in the later slides of https://www.unicodeconference.org/presentations-43/S2T2-Sheeter-Reiger.pdf.

**Comment by scottjehl** [comment]
Hi all! I'm curious if there are any updates on this. 

The main driver I see for having reliable, static font URLs on Google Fonts is that the current Google Fonts embed requires a third-party render-blocking request in order to receive the CSS that references the fonts, and that makes using google fonts rather costly to page delivery performance. Many developers are interested in avoiding this costly render-blocking request by referencing the woff2 files on google's servers directly via font-face rules that are embedded in a site's own CSS file. If we could do this reliably (knowing a given font URL will remain available over time), it would also open up the possibility of preloading those fonts when desired.

After seeing @addyosmani's second bullet above, the "latest" idea sounds attractive, but what remains unclear for me is whether or not the woff2 font URLs are already stable. While it's understood that different browsers receive font files that cater to their needs, it's possible that a full round trip to receive css that points to a specialized font variation may not be better for performance than simply starting with a font-face rule that references one woff2 version that works across all browsers (even if it's a little heavier). I don't know if this is true, but it'd be great to know if we could experiment with it.

So more specifically, would it be reasonable for a site to reference say, the hinted woff2 version from their own font-face rules today and avoid the upfront CSS request? 

Thanks!

**Comment by rsheeter** [comment]
> whether or not the woff2 font URLs are already stable

It's a bit of a theory/practice gap. In theory we do not guarrantee url stability or continued existence. I don' t think we want to commit to the urls never changing. In practice the fonts.gstatic.com/s/_family_/v_#_/ urls have been stable for a long time.

**Comment by scottjehl** [comment]
Thanks @rsheeter 

Is there any possibility to make a push for closing that gap in the interest of avoiding the performance toll of the existing loader? It'd be amazing to see copy/paste-able font-face rules as one of the embed options (ideally the default) on Google fonts, for example, as a supported approach.

**Comment by scottjehl** [comment]
(If so, I'm very happy to file an issue requesting this change and include some performance findings if helpful)

**Comment by rsheeter** [comment]
Tbh we're a bit strapped for resources so "possibility" is about as good as we can do :)

For context, I worry a bit about the loss of our ability to deploy ongoing optimizations. Unicode-range, woff2, and in future [IFT](https://w3c.github.io/IFT/Overview.html) are made far more impactful by our ability to deploy updates to users w/o action on their part. I also appreciate that people actively copy our UA-specific CSS (NextJS does iirc) and that it would be better if there was a clear way a version that is safe to reuse x-browser.

Perf findings are definitely helpful if that's not too great an inconvenience.

**Comment by valtlai** [comment]
I’ve experimented the following to fetch an unhinted or a hinted font depending on the device pixel ratio:

```css
@media (resolution >= 2x) {
  @font-face {
    font-family: My-Font;
    src: url(my-font-without-hints.woff2);
  }
}

@media (resolution < 2x) {
  @font-face {
    font-family: My-Font;
    src: url(my-font-with-hints.woff2);
  }
}
```

I’m not sure if this is a good approach or not.

**Comment by scottjehl** [comment]
@rsheeter here's a little repo of reduced test pages I spun up just to see what sort of performance opportunities are there. The readme has notes on the findings, but it might help to just skip to [the observations](https://github.com/scottjehl/google-fonts-testing?tab=readme-ov-file#observations) first. 

TLDR; looks like a font-face-first loading strategy tends to improve the FCP metric by almost 1 second on 4G over the existing Google Fonts link-based embed, regardless of whether it uses font-display: swap or default font-display behavior (this was surprising to me). All tests use a hinted (read: slightly heavier) woff2 file to represent a single woff file that can work across devices, even if it contains some overhead that some browsers won't use.

If helpful, I can file this as a new issue request that specifically asks for a font-face-based embed option. Thanks!

https://github.com/scottjehl/google-fonts-testing

**Comment by rsheeter** [comment]
Thank you for the test, most appreciated! Two things pop out to me:

1. It's requesting a full weight axis (the default) instead of just the weight it's using
   * E.g. `<link href="https://fonts.googleapis.com/css2?family=SUSE&display=swap" rel="stylesheet">` (only the default weight, 400) instead of `<link href="https://fonts.googleapis.com/css2?family=SUSE:wght@100..800&display=swap" rel="stylesheet">` (the entire weight axis)
1. There aren't very many actual pages that have no resources at all except a web font

I don't think either point changes the core result: inlining CSS can improve FCP. 

Let me think a bit about what would be involved in supporting safe-to-copy css.

**Comment by scottjehl** [comment]
Good feedback, thank you @rsheeter. Both points make sense. It'd be ideal to reference just the desirable weights.

As for point 2, I agree, but there may be a number of sites that'll see benefits from avoiding an additional render-blocking round trip in the critical path. 

Thanks for taking a look!

**Comment by rsheeter** [comment]
Thinking "aloud," I think the simplest path is to add a query param, which I'll call `mode` for now, that allows production of css based on a fixed set of capabilities (think supports woff2, variations, etc) instead of the ones for the current UA. That would enable things like site pre-processors to safely grab CSS and would make it possible for fonts.google.com to offer inline as a third option on the Embed code Web section.

For modes I'm hand wavily imagining something like:

1. `optimal` - default, best we can do for the UA. The current behavior. Not guarranteed to be safe to reuse for other UAs.
1. `modern2024` - works on all modern browsers as of 2024. Probably doesn't work entirely right on certain very old Mac browsers that don't handle overlaps in variable fonts correctly.
1. `compat2024` - as close as possible to works everywhere, even at cost of capability.
1. `mobile2024` - optimize for modern mobile browsers as of 2024. E.g. remove hints. 

Even just the first 2 options would be a good start.

Once something like incremental transfer is widely supported we might add new options, e.g. `modern2027` or similar.

**Comment by scottjehl** [comment]
This sounds fantastic, @rsheeter  
Am I understanding correctly that the first option would allow us to point to a woff2 url directly using a query param and browsers would receive an appropriately tailored font file that say, might include hinting or might not depending on the UA's needs? This sounds really great.

**Comment by rsheeter** [comment]
Sorry, poorly explained. The query would go on the css api, e.g. `/css2?family=Donkey&mode=modern2024` would produce css one could safely inline.

**Comment by scottjehl** [comment]
Ah! Thanks, that sounds workable as well. Very nice.

**Comment by D3vaLP36** [comment]
I<link rel="preload" as="font" type="font/woff2" 
  href="https://fonts.gstatic.com/s/roboto/latest" crossorigin
  min-connection=(veryfast, unmetered)>
